### PR TITLE
fix(android-auto): stop the car swallowing Next/Previous presses (#638)

### LINKED ISSUE #638: [Android Auto] Fix next/previous track controls not changing songs in v0.2.6
## Summary

In Linthra **v0.2.6**, Android Auto can show playback, but using the car's **Next / Previous track controls does not reliably change the current song**.

This is a real usability regression because Android Auto media controls must be able to drive Linthra's active playback queue without requiring the phone UI.

## Goal

Make Android Auto / Android media-session track navigation fully reliable so **Next** and **Previous** always reach Linthra's real playback queue/controller and update playback, metadata and media-session state together.

## Reproduction

1. Install/use Linthra v0.2.6 on Android.
2. Start playback from Linthra.
3. Connect to Android Auto.
4. Open Linthra's media controls in Android Auto.
5. Press **Next** or **Previous** from the car/head-unit controls.

### Current behavior

The current song may remain unchanged even though the Android Auto control was pressed.

### Expected behavior

- **Next** advances exactly once to the next valid queue item.
- **Previous** follows Linthra's existing previous-track semantics and changes playback appropriately.
- The phone UI, Android notification/media controls and Android Auto all stay synchronized.

## Investigation scope

Audit the full Android media-session command path rather than patching only the Android Auto UI.

Likely areas include:

- Android `MediaSession` / `MediaSessionService` integration
- `audio_service` / media notification callbacks if still used
- Linthra's `MediaSessionBinding` seam
- playback controller / queue controller
- next/previous command routing
- queue index synchronization
- playback-state and metadata publication to Android
- lifecycle/reconnect behavior when Android Auto attaches after playback has already started

Confirm whether the bug is caused by:

- next/previous callbacks not being registered;
- commands reaching a stale controller instance;
- commands being acknowledged by Android but not forwarded to Linthra;
- queue/media-session index desynchronization;
- Android Auto reconnect creating a stale session;
- command support/actions not being advertised correctly;
- command handling working in the notification but not through Android Auto;
- source-specific playback paths bypassing the shared controller.

Do not assume the first suspected cause is correct; trace the command end-to-end.

## Requirements

- Android Auto Next/Previous must use Linthra's existing playback/queue APIs.
- Do not create a separate Android Auto queue model.
- Do not manually mutate queue indices behind the playback controller.
- One button press must produce at most one navigation action.
- Commands must remain safe when:
  - the queue has one item;
  - the user is at the first or last item;
  - repeat/shuffle are enabled;
  - a track is unavailable and normal fallback/skip logic applies;
  - playback is paused;
  - the app is backgrounded;
  - the phone screen is locked;
  - Android Auto reconnects;
  - playback was started before Android Auto connected.
- Metadata/artwork/title shown in Android Auto must update after the track changes.
- Notification controls and headset/Bluetooth media buttons must not regress.
- Existing Android playback behavior outside Android Auto must remain unchanged.
- Linux behavior must remain unchanged.

## Previous-button semantics

Preserve Linthra's existing policy for Previous rather than inventing an Android-Auto-only rule. If Linthra currently restarts the current track after a playback-position threshold and otherwise goes to the previous queue item, Android Auto should trigger the same shared behavior.

## Tests

Add focused regression coverage around the media-session command bridge.

At minimum cover:

- Next command reaches the playback controller;
- Previous command reaches the playback controller;
- one command advances exactly once;
- current queue index updates correctly;
- metadata updates after navigation;
- commands work while playback is paused;
- reconnecting/rebinding the media session does not leave stale callbacks;
- queue boundary behavior;
- repeat/shuffle behavior remains correct;
- command support is advertised to Android while those actions are available;
- no duplicate command handling after lifecycle/reconnect events.

Where practical, add an Android integration test around the real media-session boundary in addition to pure Dart/controller tests.

## Manual Android Auto smoke test

Document and verify on a real Android Auto connection or supported emulator/test environment:

1. Start an album/playlist with at least 3 tracks.
2. Connect Android Auto.
3. Press Next several times.
4. Press Previous several times.
5. Pause, then press Next.
6. Lock the phone and retry.
7. Disconnect/reconnect Android Auto and retry.
8. Start playback before connecting Android Auto and retry.
9. Confirm title/artwork/progress update after each navigation action.
10. Confirm notification/headset Next/Previous still work afterward.

## Acceptance criteria

- [ ] Android Auto Next changes to the next track reliably.
- [ ] Android Auto Previous follows Linthra's shared previous-track behavior reliably.
- [ ] Each button press is handled exactly once.
- [ ] Queue index and actual playback remain synchronized.
- [ ] Android Auto metadata updates to the newly playing track.
- [ ] Behavior survives pause/background/lock-screen/reconnect scenarios.
- [ ] Notification and Bluetooth/headset media controls do not regress.
- [ ] Representative command-routing/lifecycle tests are added.
- [ ] A manual Android Auto smoke test passes on v0.2.6+ fix builds.

## Regression

Observed on **Linthra v0.2.6**.

This issue should be treated as an Android media-control reliability bug, not as a request for new Android Auto UI features.

### PR DESCRIPTION
**Pull request by TheZupZup**
Closes #638

## What was wrong

The command path was never broken. Next and Previous reach `LinthraAudioHandler`, which forwards them to the one shared `PlaybackController` — that part works, and the existing tests were right about it.

What was broken is the state the session reported *while* a track change was in flight.

`LinthraAudioHandler` mapped the controller's `PlaybackStatus.loading` to `AudioProcessingState.loading`. That is the one value audio_service turns into `PlaybackStateCompat.STATE_CONNECTING` (`AudioService.getPlaybackState()`), and STATE_CONNECTING does not mean "preparing the next item" — the platform defines it as the session connecting to a new destination. Android Auto reads it that way: it shows its connecting placeholder, drops the progress it was drawing, and stops dispatching transport presses.

Linthra enters `loading` on *every* track change. The controller publishes it before resolving the next track's playable URI, which on a remote source is a network round trip. So each press opened a window in which the following press was acknowledged by the car and never reached the app at all. That is "works sometimes", which is what the report said.

It also explains the one thing that looked odd about the bug: notification, Bluetooth and headset buttons never regressed. Those arrive as media-button intents, which the session dispatches whatever state it is in.

## The fix

A track change now reports `buffering`, which is also simply the truthful state — the item is being prepared, and that is what `STATE_BUFFERING` is for. The car keeps its transport row live and its metadata and progress visible across the change.

This is a reporting change only:

- the queue, the controller, and `PlaybackStatus` are untouched;
- no Android-Auto queue model, no index touched behind the controller;
- the session keeps reporting `playing` through the transition, so the foreground service is held exactly as before (#244 / #499);
- MPRIS has its own mapping, so Linux is unchanged;
- in-app playback reads the controller, so nothing outside the session moves.

Previous is unchanged in behaviour: `skipToPrevious` calls `PlaybackController.skipToPrevious` and nothing else, so the car gets Linthra's shared policy (step back, no restart-first) and keeps getting it if that policy ever changes.

## Lifecycle guards the audit turned up

None of these is the root cause, all are real and cheap:

- **`connectMediaSession` attaches at most once per process.** A second `AudioService.init` installs fresh platform callbacks and fresh stream observers without tearing the previous ones down, which leaves two handlers mirroring into the same session — two writers over one notification and one Up Next list. It is an assertion error in debug and silent in release, which is the shape of a bug that only shows up on a real head unit. Attaching is asynchronous, so an attach already in flight is *joined* rather than raced.
- **A disposed handler is inert.** It forwards nothing to the controller, so a command still in flight across a reconnect can't drive a controller it no longer speaks for. `playFromMediaId` re-checks after its resolve as well as before it — it is the one command that awaits before touching the controller.
- **A reset settles completely.** The test-only detach supersedes an attach still in flight (generation token) and awaits it, so it never returns while an `AudioService.init` is still running.

Next/Previous also breadcrumb through `StabilityDiagnostics` now. "The car's Next did nothing" has two opposite causes and only this separates them: no breadcrumb means the press never arrived, a breadcrumb with no track change means it arrived and was a legitimate queue-boundary no-op.

## Tests

`test/core/services/android_media_session_boundary_test.dart` is new and works at the real boundary rather than against Dart mocks. It crosses the actual platform channels the Android plugin uses, both ways:

- **inbound** — messages are encoded exactly as `AudioServicePlugin.invokeMethod("skipToNext", mapOf())` sends them from `MediaSessionCallback`, so the decode and dispatch path a car's press takes is the one under test;
- **outbound** — assertions read the raw argument maps the plugin would parse: the `PlaybackStateCompat` action bits, the active queue item id, the state constant.

Without the fix it fails on `processingState: 1` (STATE_CONNECTING). It also covers the media-button `click` path, so the notification/Bluetooth route is pinned against regression, and it attaches twice concurrently to prove one handler and one `configure`.

Dart-level regression coverage added alongside it: next callback, previous callback, exactly-once dispatch, a burst of presses, queue index synchronization, metadata after navigation, paused navigation, repeat and shuffle, first and last queue item, and the stale-handler and reconnect guards — including a gated browse tree that detaches the handler mid-lookup.

`flutter analyze`, `dart format --set-exit-if-changed .`, the secret/privacy scan, the Android channel threading check and the full suite (5556 tests) all pass.

## Docs

`docs/android-auto.md` gains a **track navigation smoke test** — a five-minute run covering the 13 steps from the issue (single press, rapid presses, previous, paused, screen locked, boundaries, repeat/shuffle, reconnect, playback-before-connect, Up Next row, notification/Bluetooth, back to the phone) — plus the breadcrumb reference and a "Next/Previous doesn't change the song" triage that starts by asking whether the command reached the app at all.

## Review round

Three P2 findings from the Codex review, all verified against the code and all real — they were concurrency edges in the new lifecycle guards, not in the fix itself. Fixed in 27ae3a5 and 942ebd7. Two carry tests proven to fail without the fix; the third (reset landing mid-attach) is fixed on reasoning alone, because reproducing its window would mean adding a production seam purely to test a `@visibleForTesting` helper. That is called out on the thread rather than implied to be covered.

## Not done

`pubspec.yaml` gains one **dev** dependency, `audio_service_platform_interface`. It is already in the graph under `audio_service`; the boundary test needs it directly to install the Android `AudioServicePlatform` on a Linux test host. Nothing in `lib/` imports it, and `docs/dependency-license-audit.md` is updated.

I have not changed what Next does at the end of the queue. It stays a no-op and does not wrap under repeat-all, matching the in-app button, the notification and MPRIS — changing it would change shared behaviour on every platform, which is outside this issue.

The manual Android Auto smoke test is documented but has **not** been run on a real head unit or the DHU — no car or Android device in this environment. That is the one acceptance-criteria item still open.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01J4aHSnXkWG6CqXQhLxC6ES

### CODE REVIEW

File: lib/core/services/linthra_audio_handler.dart [hunk 1]
```diff
@@ -813,6 +925,14 @@ Future<LinthraAudioHandler?> connectMediaSession(
   DownloadRepository? downloads,
   MediaArtworkSource? artwork,
 }) async {
+  final LinthraAudioHandler? attached = _attachedHandler;
+  if (attached != null && !attached.isDetached) {
```
[the remark this answers is not in the document]
Comment (TheZupZup) [reply lib/core/services/linthra_audio_handler.dart:938]: Good catch, fixed in 27ae3a5. `connectMediaSession` now holds the in-flight attach and joins concurrent callers onto it instead of letting each start its own `AudioService.init`. The slot is released when the attach settles, identity-guarded, so a failed attach can still be retried.

The boundary test now attaches twice concurrently and asserts both callers get the same handler and exactly one `configure` crossed the channel. It fails without the fix.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: lib/core/services/linthra_audio_handler.dart [hunk 2]
```diff
@@ -297,6 +353,7 @@ class LinthraAudioHandler extends audio.BaseAudioHandler {
     String mediaId, [
     Map<String, dynamic>? extras,
   ]) async {
+    if (_detached) return;
     final request = await _tree.resolve(mediaId, _controller.state);
```
[the remark this answers is not in the document]
Comment (TheZupZup) [reply lib/core/services/linthra_audio_handler.dart:357]: Right, fixed in 27ae3a5. `playFromMediaId` re-checks `_detached` after the resolve as well as before it. This was the only command with that shape — every other one reaches the controller synchronously after its check, so only the media-id lookup could be detached mid-flight.

Covered by a test that holds the browse tree's `resolve` open, disposes the handler while it is still running, and asserts the controller never starts playing. It fails without the fix.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: lib/core/services/linthra_audio_handler.dart [hunk 3]
```diff
@@ -882,3 +1049,28 @@ Future<LinthraAudioHandler?> connectMediaSession(
...
+///
+/// Only for tests. Production never detaches on Android: the session outlives
+/// the app object (see `AudioServiceMediaSessionBinding`). Disposing the old
+/// handler is the point — it makes it inert, so a stale command can't reach a
+/// controller it no longer speaks for.
+@visibleForTesting
+Future<void> resetAttachedMediaSession() async {
+  final LinthraAudioHandler? handler = _attachedHandler;
+  _attachedHandler = null;
+  _attaching = null;
```
[the remark this answers is not in the document]
Comment (TheZupZup) [reply lib/core/services/linthra_audio_handler.dart:1099]: Correct, fixed in 942ebd7 with the generation token.

A reset now bumps `_attachGeneration`; an attach that finds the generation changed while it was in flight disposes the handler it built and publishes nothing, so it can't install itself over whatever the reset left behind. The reset also awaits the superseded attach, so it never returns while an `AudioService.init` is still running — that was the part that let a following `connectMediaSession` start a second init alongside the first.

One thing worth stating plainly: this seam is `@visibleForTesting`. Production never detaches on Android, where the session outlives the app object, so the practical blast radius was state leaking between tests rather than a device-side fault. I have not added a test for it — reproducing the window needs control over when `AudioService.init` completes, which would mean adding a production seam purely to test a test-only helper. The other two findings do have tests that fail without their fix; this one is fixed on reasoning alone, and I would rather say so than imply coverage that isn't there.

---
_Generated by [Claude Code](https://claude.ai/code)_
